File: Graph/Topological Sort(DFS).c
#include <stdio.h>
#include <stdlib.h>

#define MAX_NODES 10

struct Node {
    int data;
    struct Node* next;
};

struct Node* graph[MAX_NODES];
int visited[MAX_NODES];
int numNodes;
int stack[MAX_NODES];
int top = -1;

struct Node* newNode(int node) {
    struct Node* newnode = (struct Node*)malloc(sizeof(struct Node));
    newnode->data = node;
    newnode->next = NULL;
    return newnode;
}

void addEdge(int src, int dest) {
    struct Node* newnode = newNode(dest);
    newnode->next = graph[src];
    graph[src] = newnode;
}

void push(int node) {
    stack[++top] = node;
}

int pop() {
    return stack[top--];
}

void dfs(int node) {
    visited[node] = 1;
    struct Node* temp = graph[node];
    while (temp != NULL) {
        int adjNode = temp->data;
        if (!visited[adjNode]) {
            dfs(adjNode);
        }
        temp = temp->next;
    }
    push(node); // Push node onto the stack after visiting all its neighbors
}

void topologicalSort() {
    for (int i = 0; i < numNodes; i++) {
        if (!visited[i]) {
            dfs(i);
        }
    }

    printf("Topological Sorting: ");
    while (top != -1) {
        printf("%d ", pop()); // Pop nodes from the stack to get the topological order
    }
}

int main() {
    printf("Enter the number of nodes: ");
    scanf("%d", &numNodes);
    int numEdges;
    printf("Enter the number of edges: ");
    scanf("%d", &numEdges);

    for (int i = 0; i < numNodes; i++) {
        visited[i] = 0;
        graph[i] = NULL;
    }

    printf("Enter the edges (format: node1 node2):\n");
    for (int i = 0; i < numEdges; i++) {
        int node1, node2;
        scanf("%d %d", &node1, &node2);
        addEdge(node1, node2);
    }

    topologicalSort();

    printf("\n");

    return 0;
}

File: Graph/topological sort(dfs).py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.adj_list = {v: [] for v in range(self.V)}

    def add_edge(self, src, dest):
        self.adj_list[src].append(dest)
    
    def topsort(self,visited,start,stack):
        visited[start]=True

        for neighbour in self.adj_list[start]:
            if(visited[neighbour]==0):
                self.topsort(visited,neighbour,stack)
        stack.append(start)
        
    def topsort_util(self,start):
        visited=[False]*self.V
        stack=[]
        for neighbour in range(self.V):
            if(visited[neighbour]==0):
                self.topsort(visited,neighbour,stack)
        print(stack[::-1])


# Example usage with user input:
if __name__ == "__main__":
    num_vertices = int(input("Enter the number of vertices: "))
    graph = Graph(num_vertices)
    num_edges = int(input("Enter the number of edges: "))
    print("Enter edges in the format 'source destination':")
    for _ in range(num_edges):
        src, dest = map(int, input().split())
        graph.add_edge(src, dest)

    start = int(input("Enter the starting vertex: "))

    graph.topsort_util(start)

        
